src/solvers/flattening/arrays.cpp

```diff
@@ -325,12 +325,8 @@ void arrayst::add_array_Ackermann_constraints()
           // index equality
           equal_exprt indices_equal(*i1, *i2);
 
-          if(indices_equal.op0().type()!=
-             indices_equal.op1().type())
-          {
-            indices_equal.op1().
-              make_typecast(indices_equal.op0().type());
-          }
+          indices_equal.op1() = typecast_exprt::conditional_cast(
+            indices_equal.op1(), indices_equal.op0().type());
```

Comment:
I'd put this right in the constructor:
`equal_exprt indices_equal(*i1, typecast_exprt::conditional_cast(*i2, i1->type()));`

Reply:
done